src/stage1/bigint.cpp

```diff
@@ -468,6 +468,54 @@ void bigint_min(BigInt* dest, const BigInt *op1, const BigInt *op2) {
     }
 }
 
+// TODO: audit leaks 
+// TODO: is there a better way to do this?
+void bigint_saturate(BigInt* dest, uint32_t bit_count, bool is_signed) {
+    BigInt max; 
+    if(is_signed) {
+        BigInt min;
+        bit_count -= 1;
+        bigint_init_signed(&min, -(1 << bit_count));
+        bigint_init_signed(&max,  (1 << bit_count) - 1);
+        if (bigint_cmp(dest, &min) ==  CmpLT){
+            *dest = min;
+        } else if (bigint_cmp(dest, &max) ==  CmpGT){
+            *dest = max;
+        }
+        bigint_deinit(&min);
+        bigint_deinit(&max);
+    } else {
+        bigint_init_unsigned(&max, (1 << bit_count) - 1);
+        if (bigint_cmp(dest, &max) ==  CmpGT){
+            *dest = max;
+        }
+        if (bigint_cmp_zero(dest) ==  CmpLT){
+            bigint_init_unsigned(dest, 0);
+        }
+        bigint_deinit(&max);
+    }
+}
+
```

Comment:
Instead of initializing a new integer, you could simply compute the number of bits required to store the value, and use that to decide whether to clamp the result. This should save a tiny bit of memory. See `bigint_bits_needed`, but be wary of signedness for that.